src/client-side-encryption/autoEncrypter.ts

```diff
@@ -0,0 +1,641 @@
+   */
+  init(callback: Callback<MongoClient>) {
+    if (this._bypassMongocryptdAndCryptShared || this.cryptSharedLibVersionInfo) {
+      return callback();
+    }
+    if (!this._mongocryptdManager) {
+      return callback(
+        new MongoRuntimeError(
+          'Reached impossible state: mongocryptdManager is undefined when neither bypassSpawn nor the shared lib are specified.'
+        )
+      );
+    }
+    if (!this._mongocryptdClient) {
+      return callback(
+        new MongoRuntimeError(
+          'Reached impossible state: mongocryptdClient is undefined when neither bypassSpawn nor the shared lib are specified.'
+        )
+      );
+    }
+    const _callback = (err?: AnyError, res?: MongoClient) => {
+      if (
+        err &&
+        err.message &&
+        (err.message.match(/timed out after/) || err.message.match(/ENOTFOUND/))
+      ) {
+        callback(
+          new MongoError(
+            'Unable to connect to `mongocryptd`, please make sure it is running or in your PATH for auto-spawn'
+          )
+        );
+        return;
+      }
+
+      callback(err, res);
+    };
+
+    if (this._mongocryptdManager.bypassSpawn) {
+      this._mongocryptdClient.connect().then(
+        result => {
+          return _callback(undefined, result);
+        },
+        error => {
+          _callback(error, undefined);
+        }
+      );
+      return;
+    }
+
+    this._mongocryptdManager.spawn(() => {
+      if (!this._mongocryptdClient) {
+        return callback(
+          new MongoRuntimeError(
+            'Reached impossible state: mongocryptdClient is undefined after spawning libmongocrypt.'
+          )
+        );
+      }
+      this._mongocryptdClient.connect().then(
+        result => {
+          return _callback(undefined, result);
+        },
+        error => {
+          _callback(error, undefined);
+        }
+      );
+    });
+  }
+
+  /**
+   * Cleans up the `_mongocryptdClient`, if present.
+   */
+  teardown(force: boolean, callback: Callback<void>) {
+    if (this._mongocryptdClient) {
+      this._mongocryptdClient.close(force).then(
+        result => {
+          return callback(undefined, result);
+        },
+        error => {
+          callback(error);
+        }
+      );
+    } else {
+      callback();
+    }
+  }
+
+  encrypt(ns: string, cmd: Document, callback: Callback<Document | Uint8Array>): void;
+  encrypt(
+    ns: string,
+    cmd: Document,
+    options: CommandOptions,
+    callback: Callback<Document | Uint8Array>
+  ): void;
+  /**
+   * Encrypt a command for a given namespace.
+   */
+  encrypt(
+    ns: string,
+    cmd: Document,
+    options?: CommandOptions | Callback<Document | Uint8Array>,
+    callback?: Callback<Document | Uint8Array>
+  ) {
+    if (typeof ns !== 'string') {
+      throw new TypeError('Parameter `ns` must be a string');
+    }
+
+    if (typeof cmd !== 'object') {
+      throw new TypeError('Parameter `cmd` must be an object');
+    }
```

Comment:
I decided not to clean up argument checking like this - for our public APIs in the CE object, I think they're worthwhile to keep.  Private APIs like this I think could be cleaned up, but for consistency and fewer changes I left them as-is.

Reply:
Can you clarify?  Why would we remove guards against invalid arguments?

Reply:
I think the argument validation is fine, a little unnecessary if we trust our TS. But I thought "clean up" would include making this MongoError subclass (like InvalidArgument)